### PR TITLE
Feature/local fk

### PR DESCRIPTION
**Pull request by annashamray**
Depends on https://github.com/VNG-Realisatie/vng-api-common/pull/87 

**Changes:**
1. Add FK-relations between different applications
1. Remove signals and `ZaakBesluit` and `ObjectInformatieObject` models and endpoints
1. Fix all tests to process DB-relations instead of service relations

### CODE REVIEW

**Comment by joeribekker** [comment]
@annashamray `vng-api-common==1.0.9` is now available.

File: src/openzaak/components/documenten/admin.py [hunk 1]
```diff
@@ -18,17 +18,11 @@ class EnkelvoudigInformatieObjectInline(admin.StackedInline):
...
-    queryset.update(lock='')
-
-
 @admin.register(EnkelvoudigInformatieObjectCanonical)
 class EnkelvoudigInformatieObjectCanonicalAdmin(PrivateMediaMixin, admin.ModelAdmin):
     list_display = ['__str__', 'get_not_lock_display']
-    list_display = ['__str__']
     inlines = [EnkelvoudigInformatieObjectInline, GebruiksrechtenInline]
     private_media_fields = ('inhoud',)
-    actions = [unlock]
```
Comment (joeribekker) [inline src/openzaak/components/documenten/admin.py]: Why was this removed?

File: src/openzaak/components/documenten/api/serializers.py [hunk 2]
```diff
@@ -396,51 +378,6 @@ def save(self, **kwargs):
         return self.instance
 
 
-class ObjectInformatieObjectSerializer(serializers.HyperlinkedModelSerializer):
```
Comment (joeribekker) [inline src/openzaak/components/documenten/api/serializers.py]: I think it was a mistake to delete the ObjectInformatieObject model, serializer and sync-methods.

Shouldn't they be still in place? We now removed a whole resource from the API, which is not allowed. I think the model etc should still be there. The sync methods can work on FK-level or whatever.

This keeps the API as-is and also keep functionality that will return anyway...

@sergei-maertens ?

File: src/openzaak/components/documenten/api/viewsets.py [hunk 3]
```diff
@@ -170,14 +162,6 @@ class EnkelvoudigInformatieObjectViewSet(NotificationViewSetMixin,
 
     @transaction.atomic
     def perform_destroy(self, instance):
-        if instance.canonical.objectinformatieobject_set.exists():
```
Comment (joeribekker) [inline src/openzaak/components/documenten/api/viewsets.py]: This should also be back in if we return OIO.

File: src/openzaak/components/documenten/openapi.yaml [hunk 4]
```diff
@@ -3602,12 +2998,9 @@ components:
           $ref: '#/components/schemas/Integriteit'
         informatieobjecttype:
           title: Informatieobjecttype
-          description: URL-referentie naar het INFORMATIEOBJECTTYPE (in de Catalogi
-            API).
+          description: Referentie naar het INFORMATIEOBJECTTYPE (in de Catalogi API).
           type: string
           format: uri
-          maxLength: 200
-          minLength: 1
```
Comment (joeribekker) [inline src/openzaak/components/documenten/openapi.yaml]: Should be fixed with general fix mentioned earlier.

File: src/openzaak/components/documenten/openapi.yaml [hunk 5]
```diff
@@ -3602,12 +2998,9 @@ components:
           $ref: '#/components/schemas/Integriteit'
         informatieobjecttype:
           title: Informatieobjecttype
-          description: URL-referentie naar het INFORMATIEOBJECTTYPE (in de Catalogi
-            API).
+          description: Referentie naar het INFORMATIEOBJECTTYPE (in de Catalogi API).
```
Comment (joeribekker) [inline src/openzaak/components/documenten/openapi.yaml]: Rename back please.

File: src/openzaak/components/documenten/openapi.yaml [hunk 6]
```diff
@@ -3925,12 +3318,9 @@ components:
...
```
Comment (joeribekker) [inline src/openzaak/components/documenten/openapi.yaml]: Rename back please.

File: src/openzaak/components/documenten/openapi.yaml [hunk 7]
```diff
@@ -3925,12 +3318,9 @@ components:
...
           type: string
           format: uri
-          maxLength: 200
-          minLength: 1
```
Comment (joeribekker) [inline src/openzaak/components/documenten/openapi.yaml]: Should be fixed with general fix mentioned earlier.

File: src/openzaak/components/documenten/tests/test_lock_eio.py [hunk 8]
```diff
@@ -157,11 +153,10 @@ def test_unlock_fail_incorrect_id(self):
         self.assertEqual(error['code'], 'incorrect-lock-id')
 
     def test_unlock_force(self):
-        self.autorisatie.scopes = [SCOPE_DOCUMENTEN_GEFORCEERD_UNLOCK]
-        self.autorisatie.save()
+        # self.autorisatie.scopes = [SCOPE_DOCUMENTEN_GEFORCEERD_UNLOCK]
+        # self.autorisatie.save()
```
Comment (joeribekker) [inline src/openzaak/components/documenten/tests/test_lock_eio.py]: Isn't it better to skip this test, since we still want to test force unlock based on scope authorizations.

File: src/openzaak/components/zaken/api/serializers/core.py [hunk 9]
```diff
@@ -112,7 +112,8 @@ class Meta:
         extra_kwargs = {
             'url': {
                 'validators': [
-                    ResourceValidator('Zaak', settings.ZRC_API_SPEC, get_auth=get_auth, headers={'Accept-Crs': 'EPSG:4326'})
+                    ResourceValidator('Zaak', settings.ZRC_API_SPEC, get_auth=get_auth,
+                                      headers={'Accept-Crs': 'EPSG:4326'})
```
Comment (joeribekker) [inline src/openzaak/components/zaken/api/serializers/core.py]: Why is this still a URL?

File: src/openzaak/components/zaken/api/serializers/core.py [hunk 10]
```diff
@@ -928,41 +837,7 @@ class Meta:
...
             'resultaattype': {
-                'validators': [
-                    IsImmutableValidator(),
-                    ResourceValidator('ResultaatType', settings.ZTC_API_SPEC, get_auth=get_auth)
-                ],
-            }
-        }
-
-
-class ZaakBesluitSerializer(NestedHyperlinkedModelSerializer):
```
Comment (joeribekker) [inline src/openzaak/components/zaken/api/serializers/core.py]: Another resource gone that should be in.

@sergei-maertens thoughts?!

File: src/openzaak/components/zaken/api/tests/test_zaakinformatieobjecten.py [hunk 11]
```diff
@@ -1,84 +1,53 @@
...
-    LINK_FETCHER='vng_api_common.mocks.link_fetcher_200',
-    ZDS_CLIENT_CLASS='vng_api_common.mocks.MockClient'
-)
-class ZaakInformatieObjectAPITests(ZaakInformatieObjectSyncMixin, JWTAuthMixin, APITestCase):
 
-    list_url = reverse(ZaakInformatieObject)
 
+class ZaakInformatieObjectAPITests(JWTAuthMixin, APITestCase):
+
+    list_url = reverse(ZaakInformatieObject)
```
Comment (joeribekker) [inline src/openzaak/components/zaken/api/tests/test_zaakinformatieobjecten.py]: I don't understand why this would work?

File: src/openzaak/components/zaken/openapi.yaml [hunk 12]
```diff
@@ -493,7 +493,7 @@ paths:
           format: uri
       - name: resultaattype
         in: query
-        description: URL-referentie naar het RESULTAATTYPE (in de Catalogi API).
+        description: URL naar de/het gerelateerde resultaattype
```
Comment (joeribekker) [inline src/openzaak/components/zaken/openapi.yaml]: Please change this back -- not mentioning this for all URL-fields.

File: src/openzaak/components/zaken/openapi.yaml [hunk 13]
```diff
@@ -9415,8 +8789,6 @@ components:
           description: URL-referentie naar de EIGENSCHAP (in de Catalogi API).
           type: string
           format: uri
-          maxLength: 1000
-          minLength: 1
```
Comment (joeribekker) [inline src/openzaak/components/zaken/openapi.yaml]: This should be back and hopefully with a generic solution.

File: src/openzaak/components/besluiten/openapi.yaml [hunk 14]
```diff
@@ -2622,12 +2618,9 @@ components:
           minLength: 1
         informatieobject:
           title: Informatieobject
-          description: URL-referentie naar het INFORMATIEOBJECT (in de Documenten
-            API) waarin (een deel van) het besluit beschreven is.
+          description: URL-referentie naar het INFORMATIEOBJECT.
           type: string
           format: uri
-          maxLength: 1000
-          minLength: 1
```
[the remark this answers is not in the document]
Comment (annashamray) [reply src/openzaak/components/besluiten/openapi.yaml]: I removed length restrictions from model fields because they are not shown in the schema for `HyperlinkedRelatedField`.

File: src/openzaak/components/besluiten/openapi.yaml [hunk 15]
```diff
@@ -48,15 +48,14 @@ paths:
           type: string
       - name: besluittype
         in: query
-        description: URL-referentie naar het BESLUITTYPE (in de Catalogi API).
+        description: URL naar de/het gerelateerde besluittype
```
[the remark this answers is not in the document]
Comment (annashamray) [reply src/openzaak/components/besluiten/openapi.yaml]: PR - https://github.com/VNG-Realisatie/vng-api-common/pull/88

File: src/openzaak/components/zaken/api/serializers/core.py (hunk 9 above)
Comment (annashamray) [reply src/openzaak/components/zaken/api/serializers/core.py]: I didn't change the model. It has an URL field to some `Zaak` resource and I thought It is external. 
@sergei-maertens Is it correct?

File: src/openzaak/components/zaken/models/utils.py [hunk 16]
```diff
@@ -162,21 +140,16 @@ def get_brondatum(zaak: Zaak, afleidingswijze: str, datum_kenmerk: str=None,
...
 
-        Client = import_string(settings.ZDS_CLIENT_CLASS)
-        client = Client.from_url(zaakbesluiten.first().besluit)
-        client.auth = APICredential.get_auth(zaakbesluiten.first().besluit)
-
         max_ingangsdatum = None
-        for zaakbesluit in zaakbesluiten:
-            related_besluit = client.retrieve('besluit', url=zaakbesluit.besluit)
-            ingangsdatum = datetime.strptime(related_besluit['ingangsdatum'], '%Y-%m-%d')
+        for besluit in zaakbesluiten:
```
Comment (sergei-maertens) [inline src/openzaak/components/zaken/models/utils.py]: should be replaced with a `Max("ingangsdatum")` aggregation query I think.

File: src/openzaak/components/besluiten/openapi.yaml (hunk 15 above)
[the remark this answers is not in the document]
Comment (sergei-maertens) [reply src/openzaak/components/besluiten/openapi.yaml]: Merged

File: src/openzaak/components/documenten/admin.py (hunk 1 above)
Comment (sergei-maertens) [reply src/openzaak/components/documenten/admin.py]: That's a 1.1 feature, and we're implementing 1.0.0 RC, so let's add it back because in 1.0.0 it's still relevant.

File: src/openzaak/components/besluiten/openapi.yaml (hunk 14 above)
[the remark this answers is not in the document]
Comment (annashamray) [reply src/openzaak/components/besluiten/openapi.yaml]: Let's discuss setting length constrants for `HyperlinkedRelatedField`:
1. I don't think that we should set max_length in the schema via Inspector without setting a real length validator, Because we will not really check if the client fits max_length requirements. 
1. As far as I get it I can't validate that the url doesn't exceed max_length, because validators are run after `to_internal_value` and the internal value of HyperlinkedRelatedField is a model instance, not a url.

Personally I'd left this urls in schema as they are until the loose-fk relations are implemented

File: src/openzaak/components/documenten/admin.py (hunk 1 above)
Comment (annashamray) [reply src/openzaak/components/documenten/admin.py]: `unlock` action is restored

File: src/openzaak/components/documenten/openapi.yaml (hunk 6 above)
Comment (annashamray) [reply src/openzaak/components/documenten/openapi.yaml]: done

File: src/openzaak/components/zaken/openapi.yaml (hunk 12 above)
Comment (annashamray) [reply src/openzaak/components/zaken/openapi.yaml]: done

File: src/openzaak/components/zaken/models/utils.py (hunk 16 above)
Comment (annashamray) [reply src/openzaak/components/zaken/models/utils.py]: aggregation is added to the query

File: src/openzaak/components/documenten/openapi.yaml (hunk 7 above)
Comment (annashamray) [reply src/openzaak/components/documenten/openapi.yaml]: please take a look at https://github.com/open-zaak/open-zaak/pull/51#discussion_r315630538

File: src/openzaak/components/zaken/openapi.yaml (hunk 13 above)
Comment (annashamray) [reply src/openzaak/components/zaken/openapi.yaml]: please take a look at https://github.com/open-zaak/open-zaak/pull/51#discussion_r315630538

File: src/openzaak/components/besluiten/openapi.yaml [hunk 17]
```diff
@@ -2268,18 +2266,16 @@ components:
...
-          minLength: 1
         zaak:
           title: Zaak
-          description: URL-referentie naar de ZAAK (in de Zaken API) waarvan dit besluit
+          description: Referentie naar de ZAAK (in de Zaken API) waarvan dit besluit
             uitkomst is.
           type: string
           format: uri
-          maxLength: 200
+          nullable: true
```
[the remark this answers is not in the document]
Comment (annashamray) [reply src/openzaak/components/besluiten/openapi.yaml]: please take a look at https://github.com/open-zaak/open-zaak/pull/51#discussion_r315630538

**Comment by annashamray** [comment]
@sergei-maertens plase take a look at `catalogi` schema tests. Two of them fail with the error:

"  'ObjectInformatieObjectViewSet' should either include a `queryset` attribute, or override the `get_queryset()` method."

These tests were not run before because `__init__` file was absent in `catalogi` app

File: src/openzaak/components/documenten/tests/test_lock_eio.py [hunk 18]
```diff
@@ -142,9 +141,9 @@ def test_unlock_success(self):
 
         self.assertEqual(eio.lock, '')
 
+    @skip('Current implementation is without authentication')
```
[the remark this answers is not in the document]
Comment (sergei-maertens) [reply src/openzaak/components/documenten/tests/test_lock_eio.py]: I actually have no idea what's common here - for me the _feeling_ is the reverse.

I mostly find it confusing because a lot of ZTC tests are skipped because functionality is not part of the spec yet, and it's hard to tell the difference with these skipped tests. IMO it's easier to hunt down the `expectedFailure`s instead of `skip`s

File: src/openzaak/components/zaken/api/serializers/core.py [hunk 19]
```diff
@@ -354,9 +320,9 @@ def validate(self, attrs):
         default_archiefstatus = self.instance.archiefstatus if self.instance else Archiefstatus.nog_te_archiveren
         archiefstatus = attrs.get('archiefstatus', default_archiefstatus) != Archiefstatus.nog_te_archiveren
         if archiefstatus:
-            ios = self._get_information_objects()
+            ios = [zio.informatieobject for zio in self.instance.zaakinformatieobject_set.all()]
```
[the remark this answers is not in the document]
Comment (sergei-maertens) [reply src/openzaak/components/zaken/api/serializers/core.py]: it's a tradeoff yeah

**Comment by sergei-maertens** [comment]
@annashamray thanks for the heads-up, hopefully I can get to it today

File: src/openzaak/components/documenten/api/viewsets.py (hunk 3 above)
Comment (sergei-maertens) [reply src/openzaak/components/documenten/api/viewsets.py]: I think it's checking for:
* `instance.canonical.besluitinformatieobject_set.exists()`
* `instance.canonical.zaakinformatieobject_set.exists()`

**Comment by sergei-maertens** [comment]
Still WIP because of OIO/ZIO/BIO

File: src/openzaak/components/documenten/api/viewsets.py (hunk 3 above)
Comment (annashamray) [reply src/openzaak/components/documenten/api/viewsets.py]: oh,  I didn't think of them. The validation check and related test cases are added

**Comment by sergei-maertens** [comment]
@joeribekker I'll tackle OIO in a separate pull request
